### PR TITLE
Fix StackOverflow in monad instances

### PR DESCRIPTION
**Pull request by pakoito**
After testing some of the Traverse instances, I noticed there was a common StackOverflow problem across some monads. This problem is caused by ap and map calling each other, in expectation that at least one of them is overrided. I've checked all monads and added a simple delegated map call.

### CODE REVIEW

**Comment by pakoito** [comment]
I fucked up the stacked branches. Merge only once IO is done.

**Comment by pakoito** [comment]
The error wasn't found before because of #103

File: katz/src/main/kotlin/katz/effects/IO.kt [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+sealed class IO<out A> {
+    abstract fun <B> map(f: (A) -> B): IO<B>
+
+    fun <B> flatMap(f: (A) -> IO<B>): IO<B> =
+            flatMapTotal(
+                    AndThen { a: A ->
+                        try {
+                            f(a)
+                        } catch (error: Throwable) {
```
Comment (raulraja) [inline katz/src/main/kotlin/katz/effects/IO.kt]: Do we want to catch nay Throwable or just those that are non fatal? https://github.com/scala/scala/blob/v2.9.3/src/library/scala/util/control/NonFatal.scala#L31

File: katz/src/main/kotlin/katz/effects/internal/AndThen.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    @Suppress("UNCHECKED_CAST")
+    fun error(throwable: Throwable, fe: Function1<Throwable, /* B */ *>): B =
+            try {
+                runLoop(null, throwable, false)
+            } catch (throwable: Throwable) {
+                fe(throwable) as B
+            }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun runLoop(_success: A?, _failure: Throwable?, _isSuccess: Boolean): B {
```
Comment (raulraja) [inline katz/src/main/kotlin/katz/effects/internal/AndThen.kt]: what is the meaning of the `_` convention? We don't have that anywhere else.

File: katz/src/main/kotlin/katz/effects/Duration.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/**
```
Comment (raulraja) [inline katz/src/main/kotlin/katz/effects/Duration.kt]: We should add attributions in the Copying or License files but I thought we decided not to have headers copyrights.

File: katz/src/main/kotlin/katz/effects/IO.kt (hunk 1 above)
Comment (pakoito) [inline katz/src/main/kotlin/katz/effects/IO.kt]: This diff is wrong, I made a mistake and is based off the IO branch. The IO changes are here https://github.com/FineCinnamon/Katz/pull/109

Could you please move the comments there?

**Comment by pakoito** [comment]
Deprecated, the issue has been already fixed.

**Pull request closed without merging**